### DISCUSSION TITLE
Reading a QRCODE Permission Issue - HoloLens

### DISCUSSION THREAD

**Discussion by tharakanandakumara** [StereoKit/StereoKit#163] [Q&A] [answered]
I am trying to read a QR code from a Hololens. I have used the example project in the repo. I have given permission for Microphone and Webcam in appmanifest file. But I am still getting Access denied error. This is my code. Do I have to proovide any other permissions?
```

    using StereoKit;
    using System;
    using System.Collections.Generic;
    
    using Microsoft.MixedReality.QR;
    
    
    class DemoQRCode
    {
    
	    public struct QRData
	    { 
    
		    public Pose pose;
		    public float size;
		    public string text;
    
		    public static QRData FromCode(QRCode qr)
		    {
			    QRData result = new QRData();
			    result.pose = World.FromSpatialNode(qr.SpatialGraphNodeId);
			    result.size = qr.PhysicalSideLength;
			    result.text = qr.Data == null ? "" : qr.Data;
			    return result;
		    }
	    }
	    
	     static void Main(string[] args)
        {
    
		    QRCodeWatcher watcher;
		    DateTime watcherStart;
		    Dictionary<Guid, QRData> poses = new Dictionary<Guid, QRData>();
    
    
		    var status =  QRCodeWatcher.RequestAccessAsync().Result;
		    if (status != QRCodeWatcherAccessStatus.Allowed)
			    return;
    
		    
		    watcherStart = DateTime.Now;
		    watcher = new QRCodeWatcher();
		    watcher.Added += (o, qr) => {
		    
			    if (qr.Code.LastDetectedTime > watcherStart)
				    poses.Add(qr.Code.Id, QRData.FromCode(qr.Code));
		    };
		    watcher.Updated += (o, qr) => poses[qr.Code.Id] = QRData.FromCode(qr.Code);
		    watcher.Removed += (o, qr) => poses.Remove(qr.Code.Id);
		    watcher.Start();
    
    
		    while (SK.Step(() =>
		    {
			    foreach (QRData d in poses.Values)
			    {
				    Lines.AddAxis(d.pose, d.size);
				    Text.Add(
					    d.text,
					    Matrix.TR(
						    d.pose.position + d.pose.Forward * d.size * 0.1f,
						    Quat.FromAngles(0, 0, 180) * d.pose.orientation),
					    Vec2.One * d.size,
					    TextFit.Squeeze,
					    TextAlign.XLeft | TextAlign.YTop);
			    }
		    })) ;
	    }
    
    
    }
```

**Comment by maluoi** [comment, marked as answer]
It's a little hard to say here, you're missing a few other important bits that could be causing more obscuring issues! The big one that stands out is that you haven't initialized StereoKit yet, definitely make sure that you make a call to `SK.Initialize` (and `SK.Shutdown` at the end), preferably before even starting up the QR code watcher.

You can see simple examples of StereoKit lifetime both [here](https://stereokit.net/) and [here](https://stereokit.net/Pages/Guides/Getting-Started.html). And a [slightly larger example](https://github.com/maluoi/StereoKit-PaintTutorial/blob/master/Program.cs#L26) with a bit more context around it :)

**Reply by tharakanandakumara** [reply]
Thanks. Found out I haven't initiate SK. I am able to read QR code,but found out all the positions retrieved from poses return origin. I have tried multiple QR coded. it reads the text but no luck with  position(x,y,z) which always return 0.

**Reply by maluoi** [reply]
Oh, check the StereoKit NuGet package version! I believe I have a bug in v0.3.1 that might be affecting `World.FromSpatialNode`? You can try v0.3.0, or v0.3.2-preview.1, the bug isn't present in either one of those!

**Reply by tharakanandakumara** [reply]
Thanks a lot Downgrading fixed the issue
